pkg/cmd/server/kubernetes/node/options/options.go

```diff
@@ -116,30 +117,35 @@ func ComputeKubeletFlagsAsMap(startingArgs map[string][]string, options configap
 	}
 
 	// we sometimes have different clusterdns options.  I really don't understand why
-	externalKubeClient, _, err := configapi.GetExternalKubeClient(options.MasterKubeConfig, options.MasterClientConnectionOverrides)
-	if err != nil {
-		return nil, err
+	if len(args["cluster-dns"]) == 0 || args["cluster-dns"][0] == "0.0.0.0" {
+		if externalKubeClient, _, err := configapi.GetExternalKubeClient(options.MasterKubeConfig, options.MasterClientConnectionOverrides); err == nil {
+			args["cluster-dns"] = getClusterDNS(externalKubeClient, args["cluster-dns"])
+		}
+	}
+
+	// there is a special case.  If you set `--cgroups-per-qos=false` and `--enforce-node-allocatable` is
+	// an empty string, `--enforce-node-allocatable=""` needs to be explicitly set
+	// cgroups-per-qos defaults to true
+	if cgroupArg, enforceAllocatable := args["cgroups-per-qos"], args["enforce-node-allocatable"]; len(cgroupArg) == 1 && cgroupArg[0] == "false" && len(enforceAllocatable) == 0 {
+		args["enforce-node-allocatable"] = []string{""}
 	}
-	args["cluster-dns"] = getClusterDNS(externalKubeClient, args["cluster-dns"])
 
 	return args, nil
 }
 
 func KubeletArgsMapToArgs(argsMap map[string][]string) []string {
-	args := []string{}
-	for key, value := range argsMap {
-		for _, token := range value {
-			args = append(args, fmt.Sprintf("--%s=%v", key, token))
-		}
+	var keys []string
```

Comment:
If we're doing this, mind running them through a `sets.String` to give back an ordered list?